Add a tsdb writer interface

JessicaGreben opened the pull request:
This PR adds one file from #7586 that is copy/pasted into another PR #7675 .

The change in the PR adds a tsdb writer interface and also a tsdbWriter object that implements the interface. This is nice to have so we have a simple wrapper around writing time series to tsdb blocks.


Signed-off-by: jessicagreben <jessicagrebens@gmail.com>

brian-brazil commented:
We shouldn't be adding dead code, PRs should generally make sense standalone. Maybe there's some tests that could use it?

roidelapluie commented:
Is that the same as teststorage?

roidelapluie commented:
what is the reason to extract it from above PR's?

JessicaGreben commented:
@brian-brazil while looking for a some tests that could use this code, I came across this code:
https://github.com/prometheus/prometheus/blob/master/tsdb/tsdbblockutil.go

It doesn't seem like we need both of these implementations since they are mostly the same. I will close this PR and use the tsdbblockutil code in my PR.

codesome commented:
@JessicaGreben Note that functions in tsdbblockutil require all samples to be presented beforehand. It would be nice (or even a requirement) to make them iterators for streaming samples for the backfill.

bwplotka commented:
Happy to reuse more, but we might want to add there some writer like we did in our PRs, for streaming / appender functionality indeed.

JessicaGreben commented:
@codesome @bwplotka agree about the iterator would be nice. 

Should I reopen this PR and try to combine the two? Or is there a use for both?

codesome commented:
If the writer interface is more handy for backfilling purposes (which I would believe so), I would say let's re-open and have the writer interface as is and use it inside `CreateBlock`. That way it won't be a dead code.

codesome commented:
(Which means we need to remove `CreateHead`, that is not used anywhere else I think other than `CreateBlock`)

JessicaGreben commented:
I'm using the block writer in the tsdbblockutil code, but still nothing is using the tsdbblockutil code. So I'm going to see if I can fix that by adding it to a test. Also I will add some other tests for this writer code.

codesome commented:
I think block_test.go uses it

JessicaGreben commented:
@codesome I don't see the code from tsdbblockutil.go used here https://github.com/prometheus/prometheus/blob/master/tsdb/block_test.go, but looks like it might be easy to add

codesome commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+func (w *TSDBWriter) Close() error {
+	_ = os.RemoveAll(w.tmpDir)
```
This should be in a `defer` because the head.Close() can/will have some file operations (like flushing some chunks to disk for the m-map which is in the buffer).

codesome commented on tsdb/tsdbblockutil.go:
```diff
@@ -61,30 +39,35 @@ func CreateBlock(samples []*MetricSample, dir string, mint, maxt int64, logger l
-	head, err := CreateHead(samples, chunkRange, chunkDir, logger)
+
+	w := NewTSDBWriter(logger, dir)
+	err := w.initHead(chunkRange)
```
Should the `chunkRange` be part of the `NewTSDBWriter` arguments and it calling the `initHead`? As `initHead` is an unexposed method and not a part of the interface, it cannot be used outside this package.

codesome commented on tsdb/tsdbblockutil.go:
```diff
@@ -61,30 +39,35 @@ func CreateBlock(samples []*MetricSample, dir string, mint, maxt int64, logger l
 	}
 
-	ulid, err := compactor.Write(dir, head, mint, maxt, nil)
+	err = w.Close()
```
We should have this in a `defer` after initialising `w` to avoid not cleaning up the files in case of any errors.

codesome commented on tsdb/tsdbblockutil.go:
```diff
@@ -61,30 +39,35 @@ func CreateBlock(samples []*MetricSample, dir string, mint, maxt int64, logger l
+			return "", err
+		}
+	}
+	err = app.Commit()
```
We can club this with the below line as `if err := app.Commit(); err != nil {` (same for other places where the only return value is an error)

codesome commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,132 @@
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "create leveled compactor")
+	}
+	id, err := compactor.Write(w.dir, w.head, mint, maxt, nil)
```
It should be `maxt + 1` here. The reason is this comment https://github.com/prometheus/prometheus/blob/7e1c27b853ff98c0cfdcd6b482ac0caa2403d135/tsdb/db.go#L827-L829
You can copy the same comments here.

codesome commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,132 @@
+
+// NewTSDBWriter create a new block writer.
+//
+// The returned writer accumulates all the series in memory until `Flush` is called.
```
Because only a part of data will be in memory and not all.
```suggestion
// The returned writer accumulates all the series in the Head block until `Flush` is called.
```

codesome commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,132 @@
+	"os"
+	"time"
+
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
```
Needs some rearranging

codesome commented:
Just came across this https://github.com/prometheus/prometheus/blob/b0145884d3816f8575f59b1bdf676f08206acb19/tsdb/block_test.go#L303-L349

We can get rid of this and bring back `CreateHead` in the blockutil and replace those above with the methods in the blockutil. But a small test just for the writer would also make sense.

JessicaGreben commented:
@codesome Thanks for your help and comments :smile: 

I think I addressed all your comments. When/if you have any free time, it would be great to get another review from you. 

About your last comment (re:https://github.com/prometheus/prometheus/pull/7950#issuecomment-697159780), I ended up not deleting the function since it resulted in a lot of changes. Instead I call the tsdbblockutil `CreateBlock` method from inside the block_test.go `createBlock` method to keep the function signature the same and that made the change much easier.

codesome commented on tsdb/blockwriter_test.go:
```diff
@@ -0,0 +1,63 @@
+
+	// add some series
+	app := w.Appender(ctx)
+	expectedLabelNames := "a"
+	l := labels.Labels{{Name: expectedLabelNames, Value: "b"}}
+	ts := int64(44)
+	_, err = app.Add(l, ts, 7)
+	testutil.Ok(t, err)
+	err = app.Commit()
+	testutil.Ok(t, err)
+	id, err := w.Flush(ctx)
+	testutil.Ok(t, err)
```
Instead of adding only 1 sample, let's have some loop and ingest some N series with M samples in each series.

You could refer https://github.com/prometheus/prometheus/blob/2e512c823fed9e5ee4272053ba6e4fcbcf364d69/tsdb/head_test.go#L1938-L1964 which does something similar

JessicaGreben replied:
I can add that, but I'm curious what are we testing by adding more than one series?

codesome replied:
Nothing much TBH in this case. But in general >1 series activates some other operations, like the heap management and other index calculations during compaction.

JessicaGreben replied:
If it doesn't test much, then should we add it? The test here should be testing that the blockwriter functions as intended. Does the test without this modification do that?

JessicaGreben replied:
I added a 2nd series, sounds like that might be sufficient to trigger the extra stuff you were talking about

codesome commented on tsdb/blockwriter_test.go:
```diff
@@ -0,0 +1,63 @@
+	id, err := w.Flush(ctx)
+	testutil.Ok(t, err)
+
+	// confirm block has the correct data
```
To confirm the data, we need to check the actual samples. You can look at my another comment which has some reference for doing this.

codesome reviewed (commented):
Almost there! Some nits and we need to update the test.

Additionally, we are still requiring `initHead` to be called from outside. It should be part of `NewBlockWriter`.

bwplotka reviewed (approved):
This is perfect from my side. Thanks for this @JessicaGreben :heart: 

LGTM!

brian-brazil commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+// initHead creates and initialises a new TSDB head.
+// blockSize should be in milliseconds.
```
The tsdb doesn't know anything about the units it is used with, so this comment is incorrect.

codesome replied:
This comment is invalid now anyway and should be removed as `blockSize` is now passed in `NewBlockWriter` and does not mention milliseconds.

brian-brazil commented on tsdb/blockwriter.go:
```diff
@@ -0,0 +1,129 @@
+	compactor, err := NewLeveledCompactor(ctx,
+		nil,
+		w.logger,
+		[]int64{durationToMillis(2 * time.Hour)},
```
Ditto, you're making assumptions here that don't belong in the TSDB code.

bwplotka replied:
Yea, I guess that is what was designed for TSDB. I guess this means, it should be part of some other package? :thinking:

brian-brazil replied:
I'd say pass in the value instead.

codesome replied:
We could use the `w.blockSize` here instead.

codesome commented on tsdb/blockwriter_test.go:
```diff
@@ -0,0 +1,74 @@
+
+	// Add some series.
+	app := w.Appender(ctx)
+	expectedLabelNames := []string{"a", "c"}
+	expectedValues := []string{"b", "d"}
+	l := labels.Labels{{Name: expectedLabelNames[0], Value: expectedValues[0]}}
+	ts1 := int64(44)
+	v1 := float64(7)
+	_, err = app.Add(l, ts1, v1)
+	testutil.Ok(t, err)
+	l = labels.Labels{{Name: expectedLabelNames[1], Value: expectedValues[1]}}
+	ts2 := int64(55)
+	v2 := float64(12)
+	_, err = app.Add(l, ts2, v2)
+	testutil.Ok(t, err)
+	err = app.Commit()
+	testutil.Ok(t, err)
+	id, err := w.Flush(ctx)
+	testutil.Ok(t, err)
```
We can make this a little more compact, else it's little harder to read
```suggestion
	ts1, v1 := int64(44), float64(7)
	_, err = app.Add(labels.Labels{{Name: "a", Value: "b"}}, ts1, v1)
	testutil.Ok(t, err)
	ts2, v2 := int64(55), float64(12)
	_, err = app.Add(labels.Labels{{Name: "c", Value: "d"}}, ts2, v2)
	testutil.Ok(t, err)
	testutil.Ok(t, app.Commit())
	id, err := w.Flush(ctx)
	testutil.Ok(t, err)
```

codesome commented on tsdb/blockwriter.go:
```diff
@@ -52,26 +52,29 @@ func durationToMillis(t time.Duration) int64 {
 		destinationDir: dir,
 		blockSize:      blockSize,
 	}
+	err := w.initHead()
+	if err != nil {
```
We have few other nits to fix, so just another one but not a blocker :)
```suggestion
	if err := w.initHead(); err != nil {
```

codesome reviewed (approved):
LGTM! Couple of nits, in test and the about duration which Brian mentioned

JessicaGreben commented:
@codesome thanks for your help. I addressed the final comments. Let me know what you think...hopefully we can merge :)

bwplotka reviewed (commented):
![](https://media2.giphy.com/media/dkGhBWE3SyzXW/giphy.gif)
